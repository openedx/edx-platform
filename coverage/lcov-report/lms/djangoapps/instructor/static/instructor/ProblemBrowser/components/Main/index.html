
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for lms/djangoapps/instructor/static/instructor/ProblemBrowser/components/Main</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../../../../../../prettify.css" />
    <link rel="stylesheet" href="../../../../../../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../../../../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../../../../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../../../../../../index.html">All files</a> lms/djangoapps/instructor/static/instructor/ProblemBrowser/components/Main</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">70.58% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>12/17</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>3/6</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>3/6</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">70.58% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>12/17</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="Main.jsx"><a href="Main.jsx.html">Main.jsx</a></td>
	<td data-value="70.58" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 70%"></div><div class="cover-empty" style="width: 30%"></div></div>
	</td>
	<td data-value="70.58" class="pct medium">70.58%</td>
	<td data-value="17" class="abs medium">12/17</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="6" class="abs medium">3/6</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="6" class="abs medium">3/6</td>
	<td data-value="70.58" class="pct medium">70.58%</td>
	<td data-value="17" class="abs medium">12/17</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-02-19T12:31:12.551Z
            </div>
        <script src="../../../../../../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../../../../../../sorter.js"></script>
        <script src="../../../../../../../../block-navigation.js"></script>
    </body>
</html>
    