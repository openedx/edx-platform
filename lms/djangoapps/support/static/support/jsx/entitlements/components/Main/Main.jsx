import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Button, Alert } from '@openedx/paragon';
import SearchContainer from '../Search/SearchContainer';
import EntitlementSupportTableContainer from '../Table/EntitlementSupportTableContainer';
import EntitlementFormContainer from '../EntitlementForm/container';

// eslint-disable-next-line react/function-component-definition
const Main = props => (
    <div className="entitlement-support-wrapper">
        <Alert
            variant="danger"
            onClose={props.dismissErrorMessage}
            show={!!props.errorMessage}
            dismissible
        >
            {props.errorMessage}
        </Alert>
        <h2>
            Student Support: Entitlement
        </h2>
        <MainContent
            isFormOpen={props.isFormOpen}
            ecommerceUrl={props.ecommerceUrl}
            openCreationForm={props.openCreationForm}
            entitlements={props.entitlements}
        />
    </div>
);

// eslint-disable-next-line react/function-component-definition
const MainContent = (props) => {
    if (props.isFormOpen) {
        return <EntitlementFormContainer />;
    }

    return (
        <div>
            <div className="actions">
                <SearchContainer />
                <Button
                    variant="primary"
                    onClick={props.openCreationForm}
                >
                    Create New Entitlement
                </Button>
            </div>
            {
                props.entitlements.length > 0
                    ? <EntitlementSupportTableContainer ecommerceUrl={props.ecommerceUrl} /> : null
            }
        </div>
    );
};

const mapStateToProps = state => ({
    entitlements: state.entitlements,
});

Main.propTypes = {
    errorMessage: PropTypes.string.isRequired,
    dismissErrorMessage: PropTypes.func.isRequired,
    openCreationForm: PropTypes.func.isRequired,
    ecommerceUrl: PropTypes.string.isRequired,
    isFormOpen: PropTypes.bool.isRequired,
    entitlements: PropTypes.arrayOf(PropTypes.string).isRequired,
};

MainContent.propTypes = {
    openCreationForm: PropTypes.func.isRequired,
    ecommerceUrl: PropTypes.string.isRequired,
    isFormOpen: PropTypes.bool.isRequired,
    entitlements: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default connect(mapStateToProps, null)(Main);
