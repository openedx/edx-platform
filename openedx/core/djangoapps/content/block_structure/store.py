"""
Module for the Storage of BlockStructure objects.
"""
# pylint: disable=protected-access


from logging import getLogger


from openedx.core.lib.cache_utils import zpickle, zunpickle

from . import config
from .block_structure import BlockStructureBlockData
from .exceptions import BlockStructureNotFound
from .factory import BlockStructureFactory
from .models import BlockStructureModel
from .transformer_registry import TransformerRegistry

logger = getLogger(__name__)  # pylint: disable=C0103


class BlockStructureStore:
    """
    Storage for BlockStructure objects.
    """

    def __init__(self, cache):
        """
        Arguments:
            cache (django.core.cache.backends.base.BaseCache) - The
                cache into which cacheable data of the block structure
                is to be serialized.
        """
        self._cache = cache

    def add(self, block_structure):
        """
        Stores and caches a compressed and pickled serialization of
        the given block structure.

        The data stored includes the structure's
        block relations, transformer data, and block data.

        Arguments:
            block_structure (BlockStructure) - The block structure
                that is to be cached and stored.
        """
        serialized_data = self._serialize(block_structure)

        bs_model = self._update_or_create_model(block_structure, serialized_data)
        self._add_to_cache(serialized_data, bs_model)

    def get(self, root_block_usage_key):
        """
        Deserializes and returns the block structure starting at
        root_block_usage_key, if found in the cache or storage.

        The given root_block_usage_key must equate the
        root_block_usage_key previously passed to the `add` method.

        Arguments:
            root_block_usage_key (UsageKey) - The usage_key for the
                root of the block structure that is to be retrieved
                from the store.

        Returns:
            BlockStructure - The deserialized block structure starting
            at root_block_usage_key, if found.

        Raises:
            BlockStructureNotFound if the root_block_usage_key is not
            found.
        """
        bs_model = self._get_model(root_block_usage_key)

        try:
            serialized_data = self._get_from_cache(bs_model)
        except BlockStructureNotFound:
            serialized_data = self._get_from_store(bs_model)
            self._add_to_cache(serialized_data, bs_model)

        return self._deserialize(serialized_data, root_block_usage_key)

    def delete(self, root_block_usage_key):
        """
        Deletes the block structure for the given root_block_usage_key
        from the cache and storage.

        Arguments:
            root_block_usage_key (UsageKey) - The usage_key for the root
                of the block structure that is to be removed.
        """
        bs_model = self._get_model(root_block_usage_key)
        self._cache.delete(self._encode_root_cache_key(bs_model))
        bs_model.delete()
        logger.info("BlockStructure: Deleted from cache and store; %s.", bs_model)

    def is_up_to_date(self, root_block_usage_key, modulestore):
        """
        Returns whether the data in storage for the given key is
        already up-to-date with the version in the given modulestore.
        """
        try:
            bs_model = self._get_model(root_block_usage_key)
            root_block = modulestore.get_item(root_block_usage_key)
            return self._version_data_of_model(bs_model) == self._version_data_of_block(root_block)
        except BlockStructureNotFound:
            pass

        return False

    def _get_model(self, root_block_usage_key):
        """
        Returns the model associated with the given key.
        """
        return BlockStructureModel.get(root_block_usage_key)

    def _update_or_create_model(self, block_structure, serialized_data):
        """
        Updates or creates the model for the given block_structure
        and serialized_data.
        """
        root_block = block_structure[block_structure.root_block_usage_key]
        bs_model, _ = BlockStructureModel.update_or_create(
            serialized_data,
            data_usage_key=block_structure.root_block_usage_key,
            **self._version_data_of_block(root_block)
        )
        return bs_model

    def _add_to_cache(self, serialized_data, bs_model):
        """
        Adds the given serialized_data for the given BlockStructureModel
        to the cache.
        """
        cache_key = self._encode_root_cache_key(bs_model)
        self._cache.set(cache_key, serialized_data, timeout=config.cache_timeout_in_seconds())
        logger.info("BlockStructure: Added to cache; %s, size: %d", bs_model, len(serialized_data))

    def _get_from_cache(self, bs_model):
        """
        Returns the serialized data for the given BlockStructureModel
        from the cache.
        Raises:
             BlockStructureNotFound if not found.
        """
        cache_key = self._encode_root_cache_key(bs_model)
        serialized_data = self._cache.get(cache_key)

        if not serialized_data:
            logger.info("BlockStructure: Not found in cache; %s.", bs_model)
            raise BlockStructureNotFound(bs_model.data_usage_key)
        return serialized_data

    def _get_from_store(self, bs_model):
        """
        Returns the serialized data for the given BlockStructureModel
        from storage.
        Raises:
             BlockStructureNotFound if not found.
        """
        return bs_model.get_serialized_data()

    def _serialize(self, block_structure):
        """
        Serializes the data for the given block_structure.
        """
        data_to_cache = (
            block_structure._block_relations,
            block_structure.transformer_data,
            block_structure._block_data_map,
        )
        return zpickle(data_to_cache)

    def _deserialize(self, serialized_data, root_block_usage_key):
        """
        Deserializes the given data and returns the parsed block_structure.
        """

        try:
            block_relations, transformer_data, block_data_map = zunpickle(serialized_data)
        except Exception:
            # Somehow failed to de-serialized the data, assume it's corrupt.
            bs_model = self._get_model(root_block_usage_key)
            logger.exception("BlockStructure: Failed to load data from cache for %s", bs_model)
            raise BlockStructureNotFound(bs_model.data_usage_key)  # lint-amnesty, pylint: disable=raise-missing-from

        return BlockStructureFactory.create_new(
            root_block_usage_key,
            block_relations,
            transformer_data,
            block_data_map,
        )

    @staticmethod
    def _encode_root_cache_key(bs_model):
        """
        Returns the cache key to use for the given
        BlockStructureModel.
        """
        return str(bs_model)

    @staticmethod
    def _version_data_of_block(root_block):
        """
        Returns the version-relevant data for the given block, including the
        current schema state of the Transformers and BlockStructure classes.
        """
        return dict(
            data_version=getattr(root_block, 'course_version', None),
            data_edit_timestamp=getattr(root_block, 'subtree_edited_on', None),
            transformers_schema_version=TransformerRegistry.get_write_version_hash(),
            block_structure_schema_version=str(BlockStructureBlockData.VERSION),
        )

    @staticmethod
    def _version_data_of_model(bs_model):
        """
        Returns the version-relevant data for the given BlockStructureModel.
        """
        return {
            field_name: getattr(bs_model, field_name, None)
            for field_name in BlockStructureModel.VERSION_FIELDS
        }
