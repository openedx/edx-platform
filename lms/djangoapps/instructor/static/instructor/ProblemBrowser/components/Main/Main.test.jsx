import { Provider } from 'react-redux';
import React from 'react';
import {
    render,
    screen,
    waitFor,
    act,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import store from '../../data/store';
import Main from './Main';

jest.mock('BlockBrowser/components/BlockBrowser/BlockBrowserContainer', () => {
    function MockedBlockBrowserContainer() {
        return (
            <div data-testid="mocked-block-browser-container" className="block-browser">
                Mocked BlockBrowserContainer
            </div>
        );
    }
    return MockedBlockBrowserContainer;
});

describe('ProblemBrowser Main component', () => {
    const courseId = 'testcourse';
    const problemResponsesEndpoint = '/api/problem_responses/';
    const taskStatusEndpoint = '/api/task_status/';
    const excludedBlockTypes = [];
    const reportDownloadEndpoint = '/api/report_download';
    let fetchCourseBlocksMock;
    let createProblemResponsesReportTaskMock;
    let onSelectBlockMock;

    beforeEach(() => {
        fetchCourseBlocksMock = jest.fn();
        createProblemResponsesReportTaskMock = jest.fn();
        onSelectBlockMock = jest.fn();
    });

    const renderMainComponent = (props = {}) => (
        render(
            <Provider store={store}>
                <Main
                    courseId={courseId}
                    createProblemResponsesReportTask={createProblemResponsesReportTaskMock}
                    excludeBlockTypes={excludedBlockTypes}
                    fetchCourseBlocks={fetchCourseBlocksMock}
                    problemResponsesEndpoint={problemResponsesEndpoint}
                    onSelectBlock={onSelectBlockMock}
                    selectedBlock="some-selected-block"
                    taskStatusEndpoint={taskStatusEndpoint}
                    reportDownloadEndpoint={reportDownloadEndpoint}
                    ShowBtnUi="false"
                    {...props}
                />
            </Provider>,
        )
    );

    describe('Initial rendering', () => {
        test('should match snapshot with basic parameters', () => {
            const { container } = renderMainComponent();
            expect(container).toMatchSnapshot();
        });

        test('should match snapshot with selected block', () => {
            const { container } = renderMainComponent({ selectedBlock: 'some-selected-block' });
            expect(container).toMatchSnapshot();
        });
    });

    describe('Dropdown interactions', () => {
        test('should fetch course blocks when dropdown is toggled', async () => {
            renderMainComponent();

            await userEvent.click(screen.getByText('Select a section or problem'));

            await waitFor(() => {
                expect(fetchCourseBlocksMock).toHaveBeenCalledTimes(1);
                expect(fetchCourseBlocksMock).toHaveBeenCalledWith(courseId, excludedBlockTypes);
            });
        });

        test('should display dropdown when toggled', async () => {
            renderMainComponent();
            expect(screen.queryByTestId('mocked-block-browser-container')).toBeNull();
            await act(async () => {
                await userEvent.click(screen.getByText('Select a section or problem'));
            });
            await waitFor(() => expect(
                screen.getByTestId('mocked-block-browser-container'),
            ).toHaveClass('block-browser'));
            await act(async () => {
                await userEvent.click(screen.getByText('Select a section or problem'));
            });
            await waitFor(() => expect(screen.queryByTestId('mocked-block-browser-container')).toBeNull());
        });
    });
});
