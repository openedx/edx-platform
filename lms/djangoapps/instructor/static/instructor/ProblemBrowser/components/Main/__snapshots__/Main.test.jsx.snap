// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProblemBrowser Main component Initial rendering should match snapshot with basic parameters 1`] = `
<div>
  <div
    class="problem-browser-container"
  >
    <div
      class="problem-browser"
    >
      <button
        class="btn"
        type="button"
      >
        Select a section or problem
      </button>
      <input
        disabled=""
        name="problem-location"
        type="text"
        value=""
      />
      <button
        class="btn"
        name="list-problem-responses-csv"
        type="button"
      >
        Create a report of problem responses
      </button>
      <div
        aria-live="polite"
        class="report-generation-status"
      />
    </div>
  </div>
</div>
`;

exports[`ProblemBrowser Main component Initial rendering should match snapshot with selected block 1`] = `
<div>
  <div
    class="problem-browser-container"
  >
    <div
      class="problem-browser"
    >
      <button
        class="btn"
        type="button"
      >
        Select a section or problem
      </button>
      <input
        disabled=""
        name="problem-location"
        type="text"
        value="some-selected-block"
      />
      <button
        class="btn"
        name="list-problem-responses-csv"
        type="button"
      >
        Create a report of problem responses
      </button>
      <div
        aria-live="polite"
        class="report-generation-status"
      />
    </div>
  </div>
</div>
`;
